design/service-apis.md

```diff
@@ -0,0 +1,249 @@
+# Service APIs
+
+This document outlines a design for implementing [Service APIs][1] in Contour Operator.
+
+## Goals
+
+- Manage Contour and dependent resources using Service APIs.
+
+## Non Goals
+
+- Support application-level Service APIs, i.e. xRoute. These APIs will be supported by Contour directly according to
+  Issue [2809][2].
+
+## Definitions
+
+### Service APIs
+
+Service APIs is an open source project managed by the Kubernetes SIG-NETWORK community. The project's goal is to evolve
+service networking APIs within the Kubernetes ecosystem. Service APIs provide interfaces to expose Kubernetes
+applications- Services, Ingress, and more.
+
+## Background
+
+Contour Operator manages Contour using the `contours.operator.projectcontour.io` Custom Resource Definition (CRD). The
+Contour Custom Resource (CR) provides the user interface for managing an instance of Contour and is handled by the
+Kubernetes API just like built-in objects. Contour Operator monitors the Contour CR and provides status on whether the
+actual state matches the desired state of the object. Contour Operator will continue to support the Contour CRD. Adding
+support for Service APIs will provide users an alternative approach to managing Contour.
+
+## High-Level Design
+
+Contour Operator will manage Contour using the `gatewayclasses.networking.x-k8s.io` and `gateways.networking.x-k8s.io`
+CRDs. The combination of GatewayClass and Gateway CRs will provide the user interface for managing an instance of
+Contour. Service APIs provides extension points to allow for custom resources to be linked at various layers of the API.
+This provides more granular customization at the appropriate places within the API structure. These extension points
+will be leveraged to expose Contour-specific configuration.
+
+## Detailed Design
+
+### GatewayClass
+
+GatewayClass is a cluster-scoped resource that defines a set of Gateways sharing a common behavior. A GatewayClass
+with `controller: projectcontour.io/contour-operator` will be managed by Contour Operator. There MUST be at least one
+GatewayClass defined in order to be able to have a functional Gateway. This is similar to IngressClass for Ingress and
+StorageClass for PersistentVolumes. The following is an example of a GatewayClass managed by Contour Operator:
+
+```yaml
+kind: GatewayClass
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gatewayclass
+spec:
+  controller: projectcontour.io/contour-operator
+  parametersRef:
+    name: example-contour-config
+    group: operator.projectcontour.io
+    kind: ContourConfig
```

Comment:
Can we re-use the `Contour` CRD here so that there's a common vocabulary?

Reply:
@jpeach here are a few other approaches that we can consider:

1. Use a ConfigMap to represent the Contour-specific configuration. This ConfigMap would be referenced by `gatewayclass.spec.parametersRef`.
2. Add a field to `contour.spec`, that if set, will cause the operator to NOT instantiate a Contour environment. For example:
```
apiVersion: operator.projectcontour.io/v1alpha1
kind: Contour
metadata:
  name: contour-sample
spec:
  controller: GatewayClass
```

`controller: GatewayClass` indicates the `Contour` is managed by a gatewayclass controller and will cause the operator to NOT instantiate a Contour environment, while `controller: Operator` (default) will cause the operator to to  instantiate a Contour environment.

3. Similar to option 2, but use an annotation instead of adding a field to `Contour`.

The more I think about it, the more I like option 2. Unless I hear otherwise, I'll update the design based on option 2.

Reply:
I agree with you @danehans that the `GatewayClass` is a place to describe how a `Gateway` is configured, then the `Gateway` is an instantiation of a specific `GatewayClass`.